airmozilla/popcorn/renderer.py

```diff
@@ -0,0 +1,101 @@
+import os
+import tempfile
+import time
+from uuid import uuid4
+
+import boto
+from boto.s3.key import Key
+from funfactory.urlresolvers import reverse
+from popcoder.popcoder import process_json
+
+from django.db import transaction
+from django.conf import settings
+from django.utils import timezone
+
+from airmozilla.base.utils import build_absolute_url, prepare_vidly_video_url
+from airmozilla.popcorn.models import PopcornEdit
+from airmozilla.main.models import Event, VidlySubmission
+from airmozilla.manage import vidly
+from airmozilla.uploads.models import Upload
+
+
+@transaction.atomic
+def render_edit(edit_id):
+    edit = PopcornEdit.objects.get(id=edit_id)
+    event = edit.event
+
+    filename = '%s.webm' % edit.id
+    filepath = os.path.join(tempfile.gettempdir(), filename)
+    if not (os.path.isfile(filepath) and
+            os.stat(filepath)[os.stat.st_size] > 0):
+        edit.status = PopcornEdit.STATUS_PROCESSING
+        edit.save()
+        process_json(edit.data, out=filepath)
+
+    upload_file = uuid4().hex[:13]
+    # Uploads file to s3
+    connection = boto.connect_s3(
+        settings.AWS_ACCESS_KEY_ID,
+        settings.AWS_SECRET_ACCESS_KEY,
+    )
+    bucket = connection.get_bucket('vidly_uploads')
+    video_key = Key(bucket)
+
+    video_key.key = upload_file
+    start = time.time()
+    video_key.set_contents_from_filename(filepath)
+    end = time.time()
+
+    video_url = video_key.generate_url(expires_in=0, query_auth=False)
+    video_url = prepare_vidly_video_url(video_url)
+
+    filesize = os.stat(filepath).st_size
+
+    Upload.objects.create(
+        user=edit.user,
+        url=video_url,
+        file_name=upload_file,
+        mime_type='video/webm',
+        size=filesize,
+        upload_time=int(end - start)
+    )
+
+    webhook_url = build_absolute_url(reverse('popcorn:vidly_webhook'))
+    token_protection = event.privacy != Event.PRIVACY_PUBLIC
+
+    tag, error = vidly.add_media(
+        url=video_url,
+        token_protection=token_protection,
+        hd=True,
+        notify_url=webhook_url,
+    )
+
+    VidlySubmission.objects.create(
+        event=event,
+        url=video_url,
+        tag=tag,
+        hd=True,
+        submission_error=error
+    )
+
+    # raise exception if error
+    if error:
+        raise Exception("Vidly upload failed")
```

Comment:
Perhaps instead do `raise Exception(error)` so that the error message contains information about what went wrong with vidly. Alternatively you'd have to look up the `VidlySubmission` object and read its `submission_error` thing.